### PR TITLE
copr: support filter before select

### PR DESCRIPTION
**Pull request by BusyJay**
@ngaut @qiuyesuifeng @coocood PTAL

### CODE REVIEW

File: src/server/coprocessor.rs [hunk 1]
```diff
@@ -286,132 +277,195 @@ fn is_point(range: &KeyRange) -> bool {
...
+
+fn collect_col_in_expr(cols: &mut HashMap<i64, ColumnInfo>,
+                       col_meta: &[ColumnInfo],
+                       expr: &Expr)
+                       -> Result<()> {
+    if expr.get_tp() == ExprType::ColumnRef {
+        let i = box_try!(number::decode_i64(expr.get_val()));
+        for c in col_meta {
+            if c.get_column_id() == i {
+                cols.insert(i, c.clone());
```
Comment (qiuyesuifeng) [inline src/server/coprocessor.rs]: Should we check whether have duplicated columns?

Comment (coocood) [inline src/server/coprocessor.rs]: duplicated column in expression is normal.

File: src/server/coprocessor.rs [hunk 2]
```diff
@@ -286,132 +277,195 @@ fn is_point(range: &KeyRange) -> bool {
...
+        if is_point(&range) {
+            if let None = try!(self.snap.get(&Key::from_raw(range.get_start().to_vec()))) {
+                return Ok(rows);
             }
-            let (key, _) = try!(res.pop().unwrap());
-            if range.get_end() <= &key {
-                debug!("reach end key: {:?} >= {:?}", key, range.get_end());
-                break;
+            let h = box_try!(table::decode_handle(range.get_start()));
+            if try!(self.should_skip(h)) {
```
Comment (qiuyesuifeng) [inline src/server/coprocessor.rs]: `should_skip` seems a little confused,  it looks like we skip this handle and jump to a new handle in this range.

File: src/server/coprocessor.rs [hunk 3]
```diff
@@ -286,132 +277,195 @@ fn is_point(range: &KeyRange) -> bool {
...
+        let tid = self.sel.get_table_info().get_table_id();
+        let columns = self.sel.get_table_info().get_columns();
+        let mut row = Row::new();
+        let handle = box_try!(datum::encode_value(&[Datum::I64(h)]));
+        for col in columns {
+            if col.get_pk_handle() {
+                row.mut_data().extend(handle.clone());
+            } else {
+                let col_id = col.get_column_id();
+                if self.cond_cols.contains_key(&col_id) {
```
Comment (qiuyesuifeng) [inline src/server/coprocessor.rs]: @coocood  Can u show a detail example for this?

**Comment by coocood** [comment]
LGTM

File: src/server/coprocessor.rs [hunk 4]
```diff
@@ -286,132 +277,195 @@ fn is_point(range: &KeyRange) -> bool {
...
-        let (datums, _) = box_try!(datum::decode(row.get_data()));
-        for (c, d) in columns.iter().zip(datums) {
-            eval.insert(c.get_column_id(), d);
+
+    fn get_row_by_handle(&self, h: i64) -> Result<Option<Row>> {
+        let tid = self.sel.get_table_info().get_table_id();
+        let columns = self.sel.get_table_info().get_columns();
+        let mut row = Row::new();
+        let handle = box_try!(datum::encode_value(&[Datum::I64(h)]));
+        for col in columns {
```
Comment (qiuyesuifeng) [inline src/server/coprocessor.rs]: Is it more effective when we sort columns by column id?

File: src/server/coprocessor.rs [hunk 5]
```diff
@@ -286,132 +277,195 @@ fn is_point(range: &KeyRange) -> bool {
...
+                return Ok(rows);
+            }
+            let (key, value) = try!(nk.pop().unwrap());
+            if r.get_end() <= &key {
+                debug!("reach end key: {:?} >= {:?}", key, r.get_end());
+                return Ok(rows);
+            }
+            let mut datums = box_try!(table::decode_index_key(&key));
+            let handle = if datums.len() > info.get_columns().len() {
+                datums.pop().unwrap()
```
Comment (qiuyesuifeng) [inline src/server/coprocessor.rs]: Why not directly return error but call pop here?

File: src/server/coprocessor.rs (hunk 4 above)
Comment (BusyJay) [inline src/server/coprocessor.rs]: No, here we must iterate all columns.

File: src/server/coprocessor.rs (hunk 5 above)
Comment (BusyJay) [inline src/server/coprocessor.rs]: Because handle must be in the value or appended to the key. Here we detect which case is encountered.

**Comment by qiuyesuifeng** [comment]
LGTM

**Comment by BusyJay** [comment]
@ngaut PTAL

**Pull request merged**